pkg/experiment/compactor/compaction_worker.go

```diff
@@ -292,6 +302,76 @@ func (w *Worker) startJob(ctx context.Context, job *metastorev1.CompactionJob) *
 	return job.Status
 }
 
+func (w *Worker) exportMetricsFromCompactedBlocks(ctx context.Context, compacted []*metastorev1.BlockMeta) {
+	functions := map[string]*queryv1.FunctionList{
+		// TODO:
+		// This must be richer. First, it should be split by tenant.
+		// Also, we could have functions associated to service_name
+		// while others are just collected generally no matter what
+		// service_name we handle
+		"pyroscope": {
+			Functions: []string{
+				"net/http.HandlerFunc.ServeHTTP",
+				"runtime.gcBgMarkWorker",
+			},
+		},
+		"ride-sharing-app": {
+			Functions: []string{
+				"net/http.HandlerFunc.ServeHTTP",
+				"runtime.gcBgMarkWorker",
+			},
+		},
+	}
+	for _, c := range compacted {
+		reader := query_backend.NewBlockReader(w.logger, w.storage)
+		var res, _ = reader.Invoke(ctx,
```

Comment:
This is quite an original solution. However, to my mind, the only reason we're implementing this in compactor is that we don't want to query data. I might be missing something but this is what we're actually doing here.

I propose an alternative approach:
1. We inject our component that extract data from samples to the blocks merge process [here](https://github.com/grafana/pyroscope/blob/main/pkg/experiment/query_backend/block/compaction.go#L348-L347).
2. For every label set we encounter while handling the stream (a unique `Series ID` after the `indexRewriter.rewriteRow` call): we check if it matches any of the rules (we still want to support label filters)
3. If series labels (`ProfileEntry.Labels`) match any rule, handle grouping to get the labels we want to preserve in the output time series (metrics). A conjunction of the `group by` label values and the rule name identifies the aggregator for the time series.
4. We add function/call-chain filtering later. For now we focus on the core functionality.

Basically, something very similar to our [downsampler](https://github.com/grafana/pyroscope/blob/main/pkg/phlaredb/downsample/downsample.go#L205), but simpler, as we don't need to handle samples and write to aux tables.

---

If we want to generate metrics with "weighted" values, where the values are fractions of the profile total (and we do, I believe), we need to do some more work. We need to get the *profile* total, not the *row* total, which is a single series sample set, while a profile may include multiple series. To get a profile total, we need to aggregate row totals by the `ProfileID` column: every profile has a unique ID, and as of v2, it's guaranteed, that all the profile samples will be present in the same table, regardless of their labels.

I suggest that we add it later, together with the function filtering.